File: backend/app/schemas/metrics.py
"""Pydantic schemas for metrics and hash chain."""

from __future__ import annotations

from pydantic import BaseModel


class EpochMetrics(BaseModel):
    epoch: int
    avg_reward: dict[str, float]
    survival_rate: float
    target_reached_rate: float
    avg_steps: float
    avg_energy_spent: float
    losses: dict[str, float]
    entropy: dict[str, float]
    mutual_information: dict[str, float]
    transfer_entropy: dict[str, float] | None = None
    zipf: dict[str, dict[str, float]] | None = None
    energy_roi: float | None = None
    comm_killed: bool = False
    interventions: list[dict] | None = None


class HashChainEntry(BaseModel):
    epoch: int
    hash: str
    prev_hash: str
    metrics_json: str


class HashChainVerifyResponse(BaseModel):
    valid: bool
    total_epochs: int
    first_invalid_epoch: int | None = None
    message: str

File: frontend/src/types/metrics.ts
export interface Trajectory {
  grid_size: number;
  target: [number, number];
  obstacles: [number, number][];
  steps: {
    A: [number, number];
    B: [number, number];
    C: [number, number];
    energy: Record<string, number>;
  }[];
}

export interface EpochMetrics {
  epoch: number;
  avg_reward: Record<string, number>;
  survival_rate: number;
  target_reached_rate: number;
  avg_steps: number;
  avg_energy_spent: number;
  losses: Record<string, number>;
  entropy: Record<string, number>;
  mutual_information: Record<string, number>;
  transfer_entropy?: Record<string, number>;
  zipf?: Record<string, ZipfResult>;
  energy_roi?: number;
  comm_killed: boolean;
  interventions?: Intervention[];
  trajectory?: Trajectory | null;
}

export interface ZipfResult {
  alpha: number;
  ks_statistic: number;
  ks_pvalue: number;
  r_squared: number;
}

export interface Intervention {
  epoch: number;
  type: string;
  params: Record<string, unknown>;
  active: boolean;
}

export interface HashChainEntry {
  epoch: number;
  hash: string;
  prev_hash: string;
  metrics_json: string;
}

export interface HashChainVerifyResult {
  valid: boolean;
  total_epochs: number;
  first_invalid_epoch: number | null;
  message: string;
}
